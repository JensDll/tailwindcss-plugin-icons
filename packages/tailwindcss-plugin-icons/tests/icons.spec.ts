import path from 'node:path'

import type { CSSRuleObject, PluginAPI } from 'tailwindcss/types/config'
import type { Mocked } from 'vitest'

import type { ColorFunction } from '~tailwindcss-plugin-icons/css'
import {
  Icons,
  SCALE,
  type ScaleFactory,
} from '~tailwindcss-plugin-icons/index'

const consoleErrorMock = vi.spyOn(console, 'error')

vi.mock('~tailwindcss-plugin-icons/cache', () => {
  return {
    IconifyFileCache: class {},
  }
})

const mockPluginApi: Mocked<PluginAPI> = {
  addUtilities: vi.fn(),
  matchUtilities: vi.fn(),
  addComponents: vi.fn(),
  matchComponents: vi.fn(),
  addBase: vi.fn(),
  addVariant: vi.fn(),
  theme: vi.fn() as never,
  config: vi.fn() as never,
  corePlugins: vi.fn(),
  matchVariant: vi.fn() as never,
  e: vi.fn(),
}

const location = path.resolve(__dirname, '__fixtures__/icons.json')

afterEach(() => {
  expect(mockPluginApi.addComponents.mock.lastCall).toMatchSnapshot()
  expect(mockPluginApi.matchComponents.mock.lastCall).toMatchSnapshot()

  if (mockPluginApi.matchComponents.mock.lastCall === undefined) {
    return
  }

  const snapshot: Record<string, CSSRuleObject | null> = {}

  for (const [iconClassName, colorFunction] of Object.entries<ColorFunction>(
    mockPluginApi.matchComponents.mock.lastCall[0],
  )) {
    snapshot[`${iconClassName}-red`] = colorFunction('red', { modifier: null })
    snapshot[`${iconClassName}-green`] = colorFunction('green', {
      modifier: null,
    })
  }

  expect(snapshot).toMatchSnapshot()
})

test("fail if icon doesn't exist", () => {
  Icons(() => {
    return {
      testIcons: {
        icons: {
          doesNotExist: {},
        },
        location,
      },
    }
  }).handler(mockPluginApi)

  expect(mockPluginApi.addComponents).toBeCalledTimes(0)
  expect(mockPluginApi.matchComponents).toBeCalledTimes(0)
  expect(consoleErrorMock).toBeCalledTimes(1)
})

test('use addComponent when not forced otherwise', () => {
  Icons(() => {
    return {
      testIcons: {
        icons: {
          colored: {},
          colorless: {},
        },
        location,
      },
    }
  }).handler(mockPluginApi)

  expect(mockPluginApi.addComponents).toBeCalledTimes(1)
  expect(mockPluginApi.matchComponents).toBeCalledTimes(1)
})

test('use matchComponents when forced as bg', () => {
  Icons(() => {
    return {
      testIcons: {
        icons: {
          'colored?bg': {},
          'colorless?bg': {},
        },
        location,
      },
    }
  }).handler(mockPluginApi)

  expect(mockPluginApi.addComponents).toBeCalledTimes(1)
  expect(mockPluginApi.matchComponents).toBeCalledTimes(1)
})

describe('scale width and height', () => {
  test('when passed per icon set', () => {
    Icons(() => {
      return {
        testIcons: {
          icons: {
            colored: {},
            colorless: {},
          },
          scale: 1.5,
          location,
        },
      }
    }).handler(mockPluginApi)

    expect(mockPluginApi.addComponents).toBeCalledTimes(1)
    expect(mockPluginApi.matchComponents).toBeCalledTimes(1)
  })

  test('when passed per icon set as function', () => {
    const scale: ScaleFactory = vi.fn(() => 42)

    Icons(() => {
      return {
        testIcons: {
          icons: {
            colored: {
              [SCALE]: 2,
            },
            colorless: {},
          },
          scale,
          location,
        },
      }
    }).handler(mockPluginApi)

    expect(scale).toBeCalledTimes(1)
    expect(scale).toBeCalledWith('colorless')
    expect(mockPluginApi.addComponents).toBeCalledTimes(1)
    expect(mockPluginApi.matchComponents).toBeCalledTimes(1)
  })

  test('when passed per icon', () => {
    Icons(() => {
      return {
        testIcons: {
          icons: {
            colored: {
              [SCALE]: 2,
            },
            colorless: {
              [SCALE]: 3,
            },
          },
          scale: 1,
          location,
        },
      }
    }).handler(mockPluginApi)

    expect(mockPluginApi.addComponents).toBeCalledTimes(1)
    expect(mockPluginApi.matchComponents).toBeCalledTimes(1)
  })
})

describe('include all', () => {
  afterEach(() => {
    expect(mockPluginApi.addComponents).toBeCalledTimes(1)
    expect(mockPluginApi.matchComponents).toBeCalledTimes(1)
  })

  test('without icons', () => {
    Icons(() => {
      return {
        testIcons: {
          includeAll: true,
          location,
        },
      }
    }).handler(mockPluginApi)
  })

  test('with default CSS', () => {
    Icons(() => {
      return {
        testIcons: {
          icons: {
            colorless: {
              color: 'red',
            },
          },
          includeAll: true,
          location,
        },
      }
    }).handler(mockPluginApi)
  })

  test('when forced as mask', () => {
    Icons(() => {
      return {
        testIcons: {
          icons: {
            'colored?mask': {
              color: 'red',
            },
          },
          includeAll: true,
          location,
        },
      }
    }).handler(mockPluginApi)
  })

  test('when forced as bg', () => {
    Icons(() => {
      return {
        testIcons: {
          icons: {
            'colored?bg': {
              color: 'red',
            },
          },
          includeAll: true,
          location,
        },
      }
    }).handler(mockPluginApi)
  })
})
