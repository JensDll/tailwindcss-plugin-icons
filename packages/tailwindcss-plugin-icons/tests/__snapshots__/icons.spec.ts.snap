// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fail if icon doesn't exist 1`] = `undefined`;

exports[`fail if icon doesn't exist 2`] = `undefined`;

exports[`include all > when forced as bg 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colored-h-flip": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='translate(24 0) scale(-1 1)'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/g%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
    ".i-test-icons-colorless-rotate": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='rotate(270 12 12)'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/g%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
  },
]
`;

exports[`include all > when forced as bg 2`] = `
[
  {
    "bg-test-icons-colored": [Function],
  },
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`include all > when forced as bg 3`] = `
{
  "bg-test-icons-colored-green": {
    "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
    "background": "var(--tw-plugin-icons-url) no-repeat",
    "backgroundSize": "100% 100%",
    "color": "red",
    "height": "1em",
    "width": "1em",
  },
  "bg-test-icons-colored-red": {
    "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
    "background": "var(--tw-plugin-icons-url) no-repeat",
    "backgroundSize": "100% 100%",
    "color": "red",
    "height": "1em",
    "width": "1em",
  },
}
`;

exports[`include all > when forced as mask 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "color": "red",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
    ".i-test-icons-colored-h-flip": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='translate(24 0) scale(-1 1)'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/g%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
    ".i-test-icons-colorless-rotate": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='rotate(270 12 12)'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/g%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
  },
]
`;

exports[`include all > when forced as mask 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`include all > when forced as mask 3`] = `{}`;

exports[`include all > with default CSS 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colored-h-flip": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='translate(24 0) scale(-1 1)'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/g%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "color": "red",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
    ".i-test-icons-colorless-rotate": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='rotate(270 12 12)'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/g%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
  },
]
`;

exports[`include all > with default CSS 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`include all > with default CSS 3`] = `{}`;

exports[`include all > without icons 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colored-h-flip": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='translate(24 0) scale(-1 1)'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/g%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
    ".i-test-icons-colorless-rotate": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cg transform='rotate(270 12 12)'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/g%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
  },
]
`;

exports[`include all > without icons 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`include all > without icons 3`] = `{}`;

exports[`scale width and height > when passed per icon 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "2em",
      "width": "2em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "3em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "3em",
    },
  },
]
`;

exports[`scale width and height > when passed per icon 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`scale width and height > when passed per icon 3`] = `{}`;

exports[`scale width and height > when passed per icon set 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1.5em",
      "width": "1.5em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1.5em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1.5em",
    },
  },
]
`;

exports[`scale width and height > when passed per icon set 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`scale width and height > when passed per icon set 3`] = `{}`;

exports[`scale width and height > when passed per icon set as function 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "2em",
      "width": "2em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "42em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "42em",
    },
  },
]
`;

exports[`scale width and height > when passed per icon set as function 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`scale width and height > when passed per icon set as function 3`] = `{}`;

exports[`use addComponent when not forced otherwise 1`] = `
[
  {
    ".i-test-icons-colored": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
      "background": "var(--tw-plugin-icons-url) no-repeat",
      "backgroundSize": "100% 100%",
      "height": "1em",
      "width": "1em",
    },
    ".i-test-icons-colorless": {
      "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='currentColor'/%3E%3C/svg%3E\\")",
      "backgroundColor": "currentColor",
      "height": "1em",
      "mask": "var(--tw-plugin-icons-url) no-repeat",
      "maskSize": "100% 100%",
      "width": "1em",
    },
  },
]
`;

exports[`use addComponent when not forced otherwise 2`] = `
[
  {},
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`use addComponent when not forced otherwise 3`] = `{}`;

exports[`use matchComponents when forced as bg 1`] = `
[
  {},
]
`;

exports[`use matchComponents when forced as bg 2`] = `
[
  {
    "bg-test-icons-colored": [Function],
    "bg-test-icons-colorless": [Function],
  },
  {
    "type": [
      "color",
      "any",
    ],
    "values": {},
  },
]
`;

exports[`use matchComponents when forced as bg 3`] = `
{
  "bg-test-icons-colored-green": {
    "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
    "background": "var(--tw-plugin-icons-url) no-repeat",
    "backgroundSize": "100% 100%",
    "height": "1em",
    "width": "1em",
  },
  "bg-test-icons-colored-red": {
    "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='%23fff'/%3E%3C/svg%3E\\")",
    "background": "var(--tw-plugin-icons-url) no-repeat",
    "backgroundSize": "100% 100%",
    "height": "1em",
    "width": "1em",
  },
  "bg-test-icons-colorless-green": {
    "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='green'/%3E%3C/svg%3E\\")",
    "background": "var(--tw-plugin-icons-url) no-repeat",
    "backgroundSize": "100% 100%",
    "height": "1em",
    "width": "1em",
  },
  "bg-test-icons-colorless-red": {
    "--tw-plugin-icons-url": "url(\\"data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath d='1' fill='red'/%3E%3C/svg%3E\\")",
    "background": "var(--tw-plugin-icons-url) no-repeat",
    "backgroundSize": "100% 100%",
    "height": "1em",
    "width": "1em",
  },
}
`;
